contracts/bridge/MiniChefV21.sol

```diff
@@ -0,0 +1,444 @@
+        address indexed to
+    );
+    event Withdraw(
+        address indexed user,
+        uint256 indexed pid,
+        uint256 amount,
+        address indexed to
+    );
+    event EmergencyWithdraw(
+        address indexed user,
+        uint256 indexed pid,
+        uint256 amount,
+        address indexed to
+    );
+    event Harvest(address indexed user, uint256 indexed pid, uint256 amount);
+    event LogPoolAddition(
+        uint256 indexed pid,
+        uint256 allocPoint,
+        IERC20 indexed lpToken,
+        IRewarder indexed rewarder
+    );
+    event LogSetPool(
+        uint256 indexed pid,
+        uint256 allocPoint,
+        IRewarder indexed rewarder,
+        bool overwrite
+    );
+    event LogUpdatePool(
+        uint256 indexed pid,
+        uint64 lastRewardTime,
+        uint256 lpSupply,
+        uint256 accSynapsePerShare
+    );
+    event LogSynapsePerSecond(uint256 synapsePerSecond);
+
+    /// @param _synapse The SYNAPSE token contract address.
+    constructor(IERC20 _synapse) public {
+        SYNAPSE = _synapse;
+    }
+
+    /// @notice Returns the number of MCV2 pools.
+    function poolLength() public view returns (uint256 pools) {
+        pools = poolInfo.length;
+    }
+
+    /// @notice Add a new LP to the pool. Can only be called by the owner.
+    /// Pools are updated before adding a new one, so the old rewards don't get messed up
+    /// @param allocPoint AP of the new pool.
+    /// @param _lpToken Address of the LP ERC-20 token.
+    /// @param _rewarder Address of the rewarder delegate.
+    function add(
+        uint256 allocPoint,
+        IERC20 _lpToken,
+        IRewarder _rewarder
+    ) external onlyOwner {
+        require(addedTokens[address(_lpToken)] == false, "Token already added");
+        for (uint256 pid = 0; pid < poolInfo.length; ++pid) {
+            updatePool(pid);
+        }
+        totalAllocPoint = totalAllocPoint.add(allocPoint);
+        lpToken.push(_lpToken);
+        rewarder.push(_rewarder);
+
+        poolInfo.push(
+            PoolInfo({
+                allocPoint: allocPoint.to64(),
+                lastRewardTime: block.timestamp.to64(),
+                accSynapsePerShare: 0
+            })
+        );
+        addedTokens[address(_lpToken)] = true;
+        emit LogPoolAddition(
+            lpToken.length.sub(1),
+            allocPoint,
+            _lpToken,
+            _rewarder
+        );
+    }
+
+    /// @notice Update the given pool's SYNAPSE allocation point and `IRewarder` contract. Can only be called by the owner.
+    /// @param _pid The index of the pool. See `poolInfo`.
+    /// @param _allocPoint New AP of the pool.
+    /// @param _rewarder Address of the rewarder delegate.
+    /// @param overwrite True if _rewarder should be `set`. Otherwise `_rewarder` is ignored.
+    function set(
+        uint256 _pid,
+        uint256 _allocPoint,
+        IRewarder _rewarder,
+        bool overwrite
+    ) external onlyOwner {
```

Comment:
@trajan0x @ChiTimesChi should probably add an array out of bounds check here; did so in #90 with [8fc88e9](https://github.com/synapsecns/synapse-contracts/pull/90/commits/8fc88e98ab9d783373cfbd1da85d84ab29bff8a2)